extensions/amp-ad-network-base/0.1/amp-ad-network-base.js

```diff
@@ -0,0 +1,238 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  FailureType,
+  RecoveryModeType,
+} from './amp-ad-type-defs';
+import {Services} from '../../../src/services';
+import {dev} from '../../../src/log';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {map} from '../../../src/utils/object';
+import {sendXhrRequest} from './amp-ad-utils';
+
+const TAG = 'amp-ad-network-base';
+
+/**
+ * @abstract
+ */
+export class AmpAdNetworkBase extends AMP.BaseElement {
+
+  constructor(element) {
+    super(element);
+
+    /** @private {?Promise<!../../../src/service/xhr-impl.FetchResponse>} */
+    this.adPromise_ = null;
+
+    /** @private {Object<string, !./amp-ad-type-defs.Validator>} */
+    this.validators_ = map();
+
+    /** @private {Object<string, !./amp-ad-type-defs.Renderer>} */
+    this.renderers_ = map();
+
+    /** @private {Object<string, string>} */
+    this.recoveryModes_ = map();
+
+    /** @private {?./amp-ad-type-defs.LayoutInfoDef} */
+    this.initialSize_ = null;
+
+    /** @const @private {!Object} */
+    this.context_ = {
+      element,
```

Comment:
let's remove the `element` and `win` for now. 
instead of giving requestUrlGenerator and validator the direct access of element and window,
we should prepare data for them.

Reply:
Done.